pkg/kubectl/cmd/convert.go

```diff
@@ -0,0 +1,158 @@
+of version specified by --output-version flag. If target version is not specified or
+not supported, convert to latest version.
+
+The default output will be printed to stdout in YAML format. One can use -o option
+to change to output destination.
+`
+	convert_example = `# Convert 'pod.yaml' to latest version and print to stdout.
+$ kubectl convert -f pod.yaml
+
+# Convert the live state of the resource specified by 'pod.yaml' to the latest version
+# and print to stdout in json format.
+$ kubectl convert -f pod.yaml --local -o json
+
+# Convert all files under current directory to latest version and create them all.
+$ kubectl convert -f . | kubectl create -f -
+`
+)
+
+// NewCmdConvert creates a command object for the generic "convert" action, which
+// translates the config file into a given version.
+func NewCmdConvert(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ConvertOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "convert -f FILENAME",
+		Short:   "Convert config files between different API versions",
+		Long:    convert_long,
+		Example: convert_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := options.Complete(f, out, cmd, args)
+			cmdutil.CheckErr(err)
+			err = options.RunConvert()
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	usage := "Filename, directory, or URL to file to need to get converted."
+	kubectl.AddJsonFilenameFlag(cmd, &options.filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().BoolVar(&options.local, "local", true, "If true, convert will NOT try to contact api-server but run locally.")
+
+	return cmd
+}
+
+// ConvertOptions have the data required to perform the convert operation
+type ConvertOptions struct {
+	builder   *resource.Builder
+	filenames []string
+	local     bool
+
+	out     io.Writer
+	printer kubectl.ResourcePrinter
+
+	outputVersion string
+}
+
+// Complete collects information required to run Convert command from command line.
+func (o *ConvertOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) (err error) {
+	o.outputVersion = cmdutil.OutputVersion(cmd, latest.GroupOrDie("").Version)
+	if !registered.IsRegisteredAPIVersion(o.outputVersion) {
+		cmdutil.UsageError(cmd, "'%s' is not a registered version.", o.outputVersion)
+	}
+
+	// build the builder
+	mapper, typer := f.Object()
+	if o.local {
+		fmt.Fprintln(out, "running in local mode...")
+		o.builder = resource.NewBuilder(mapper, typer, f.NilClientMapperForCommand())
+	} else {
+		o.builder = resource.NewBuilder(mapper, typer, f.ClientMapperForCommand())
+		schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+		if err != nil {
+			return err
+		}
+		cmdNamespace, _, err := f.DefaultNamespace()
+		if err != nil {
+			return err
+		}
+		o.builder = o.builder.Schema(schema).
+			NamespaceParam(cmdNamespace).DefaultNamespace()
```

Comment:
This is for `kubectl convert -f pod.yaml | kubectl create -f -` 

If enforceNamespace is true, converting spec with different namespace is forbidden. But since `convert` doesn't affect the live state of resources, we don't need this limitation, so we set enforceNamespace to false. The conversion is allowed.

Consider the senario:
User A can access namespace N
User A can't access namespace M
When spec to be converted is using namespace M, we should not prevent users from converting it. But if users want to use the converted spec, they need to modify the namespace from M to N. `DefaultNamespace` did this for users.

Does that sound resonable to you?